### PR TITLE
[build-script] Move non-trivial argument default logic

### PR DESCRIPTION
**Pull request by Rostepher**
# Purpose

This PR moves some argument default logic which was originally relocated into the `build_swift` package along-side the rest of the argument parser logic. However, upon further reflection these non-trivial default settings ought be moved back into `build-script` until they can find a more suitable home in the (near) future. The argument parser should not be handling complex defaults or inferences for values.

### CODE REVIEW

**Comment by jrose-apple** [comment]
Tiny request: can you include "build-script" in your PR titles in the future? As is it's kind of generic, and I had to click in to find out what it was about.

**Comment by Rostepher** [comment]
@jrose-apple Sure thing.

**Comment by gottesmm** [comment]
1d94cd0 needs to be split up. You are doing a bunch of other refactoring there.

File: utils/build_swift/tests/test_driver_arguments.py [hunk 1]
```diff
@@ -323,217 +419,255 @@ def setUp(self):
...
+            self.assertFalse(namespace.test_android_host)
 
     def test_implied_defaults_host_test(self):
         with self.assertNotRaises(ParserError):
-            args = self.parse_args(['--host-test', '0'])
-            self.assertFalse(args.test_ios_host)
-            self.assertFalse(args.test_tvos_host)
-            self.assertFalse(args.test_watchos_host)
-            self.assertFalse(args.test_android_host)
+            namespace = self.parse_default_args(['--host-test', '0'])
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: Specifically, this stuff. Can you put this in a different commit?

File: utils/build_swift/tests/test_driver_arguments.py [hunk 2]
```diff
@@ -323,217 +419,255 @@ def setUp(self):
...
     # Manual option tests
 
     def test_option_clang_compiler_version(self):
         option_string = '--clang-compiler-version'
 
         with self.assertNotRaises(ParserError):
-            self.parse_args([option_string, '5.0.0'])
-            self.parse_args([option_string, '5.0.1'])
-            self.parse_args([option_string, '5.0.0.1'])
+            self.parse_default_args([option_string, '5.0.0'])
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: This should be in a separate commit.

File: utils/build_swift/tests/test_driver_arguments.py [hunk 3]
```diff
@@ -323,217 +419,255 @@ def setUp(self):
...
+
+        expected_options = {o.option_string for o in eo.EXPECTED_OPTIONS}
+
+        # aggregate and flatten the options_strings accepted by the parser
+        actual_options = [a.option_strings for a in self.parser._actions]
+        actual_options = set(sum(actual_options, []))
+
+        diff = actual_options - expected_options
+
+        if len(diff) > 0:
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: Can you invert this if condition? (i.e.:

```
if len(diff) <= 0:
    return
self.fail( ...
```

File: utils/build_swift/tests/test_driver_arguments.py [hunk 4]
```diff
@@ -244,40 +315,39 @@ def test(self):
...
-        elif option.__class__ is expected_options.ToggleOption:
-            return cls._generate_toggle_option_test(option)
-        elif option.__class__ is expected_options.UnsupportedOption:
-            return cls._generate_unsupported_option_test(option)
-
-        # Ignore all IgnoreOption tests since they should be manually tested
-        elif option.__class__ is expected_options.IgnoreOption:
-            return lambda self: None
+        generate_test_funcs = {
+            eo.HelpOption: cls._generate_help_option_test,
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: The refactoring part of this as well.

**Review by gottesmm — COMMENTED** [review]
Some examples of what I am talking about.

**Comment by Rostepher** [comment]
I've rebased onto master, so the old changes that had snuck in are no longer in this PR.

**Comment by Rostepher** [comment]
@gottesmm Those test changes were from an older PR which has since been merged.

**Comment by Rostepher** [comment]
@swift-ci please smoke test

**Comment by gottesmm** [comment]
@Rostepher What is this new home you speak of? Can you give me the long term vision?

File: utils/build_swift/tests/expected_options.py [hunk 5]
```diff
@@ -69,7 +69,7 @@
     'build_playgroundsupport': False,
     'build_runtime_with_host_compiler': False,
     'build_stdlib_deployment_targets': ['all'],
-    'build_subdir': 'Ninja-DebugAssert',
```
Comment (zisko) [inline utils/build_swift/tests/expected_options.py]: Why is this default gone?

File: utils/build_swift/tests/expected_options.py [hunk 6]
```diff
@@ -145,18 +145,7 @@
...
-        'iphonesimulator-i386',
-        'iphonesimulator-x86_64',
-        'appletvsimulator-x86_64',
-        'watchsimulator-i386',
-        'iphoneos-armv7',
-        'iphoneos-armv7s',
-        'iphoneos-arm64',
-        'appletvos-arm64',
-        'watchos-armv7k'
-    ],
```
Comment (zisko) [inline utils/build_swift/tests/expected_options.py]: The idea here is that we're taking this out of the argument parser and inferring it ourselves if there is nothing passed, right?

**Review by zisko — APPROVED** [review]
It looks like we're just removing dumb defaults and inferring them in a smarter way. I like this change :)

File: utils/build_swift/tests/expected_options.py (hunk 5 above)
Comment (zisko) [reply utils/build_swift/tests/expected_options.py]: nevermind, I see that we're computing this if it's not passed.

File: utils/build_swift/tests/expected_options.py (hunk 6 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: Exactly. Rather than inferring the default stdlib deployment targets as a part of the argument parsing logic it makes more sense to set the default value to `None` if the user didn't explicitly pass the flag and infer the default value in a more suitable location.

File: utils/build_swift/tests/expected_options.py (hunk 5 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: I've also moved the `build_subdir` inference logic back to `build-script` which means that the default value will be `None` and we'll infer it.

**Comment by Rostepher** [comment]
@gottesmm My overarching plan for build-script and build-script-impl is really to consolidate their logic into the `build_swift` package with a lot of unit-tests (and maybe integration tests). `build-script-impl` will disappear in the process leaving just `build-script`, which will become a thin wrapper around functionality in `build_swift`.

Stepping in that direction will require a lot of iterative changes to move portions of functionality from `build-script`, `build-script-impl` and `swift_build_support` into `build_swift` adding tests as we go. Once the functionality has been moved it can be refactored into a more maintainable and cleaner abstraction.

The more suitable place in this context would be closer to the top-level of the script. We should not be inferring argument default values with 'non-trivial' logic as a post-processing step to the argument parsing, rather we should opt to use `None` in more places as a default value and infer suitable values in the main function.

We can have a more lengthly discussion at a later time, either in person or via mailing list. For the time being I'd like to move this PR along and wrap up my changes to the argument parsing.

**Comment by shahmishal** [comment]
We should run full testing, with this major change in build system

**Comment by Rostepher** [comment]
In the future I'll make sure to run full test on all build system changes.
